tactics/tactics.ml

```diff
@@ -315,6 +315,36 @@ let convert_concl ~cast ~check ty k =
     end
   end
 
+let with_reverse_casts env env' sigma c =
+  let ev, l = EConstr.destEvar sigma c in
+  let () = match l with
+    | Nil | Default (_, Nil) -> ()
+    | _ -> assert false
+  in
+  let rec remake_instance acc k = function
+    | [] -> acc
+    | false :: rest -> remake_instance (SList.default acc) k rest
+    | true :: rest -> remake_instance (SList.cons (mkRel k) acc) (k-1) rest
```

Comment:
Based on the error messages, it looks like there might be a de Bruijn indexing bug somewhere.  Are evar instance variables actually numbered 0,-1,-2, etc?  (Am I misinterpreting this code?)

Reply:
k doesn't start at 0, we should be producing 3, 2, 1 with defaults interleaved.

Reply:
Ah, nevermind, I think I was misreading

Reply:
if `env` has `H : t` and `env'` has `H : t'` we generate `let H : t' := H (* Var *) :- t' in ?e@{H := H (* Rel *)}`